File: 129/javascript/ch-1.js
#!/usr/bin/env node

/**
 * Task 1 > Root Distance
 * ======================
 *
 * Given a tree and a node of a given tree, write a script to find out the
 * distance of the node from the root.
 *
 * Example 1
 * =========
 *
 * Tree:
 *       1
 *      / \
 *     2   3
 *          \
 *           4
 *          / \
 *         5   6
 *
 * Node: 6
 * Output: 3 as the distance of given node 6 from the root (1).
 *
 * Node: 5
 * Output: 3
 *
 * Node: 2
 * Output: 1
 *
 * Node: 4
 * Output: 2
 *
 * Example 2
 * =========
 *
 * Tree:
 *       1
 *      / \
 *     2   3
 *    /     \
 *   4       5
 *    \     /
 *     6   7
 *    / \
 *   8   9
 *
 * Node: 7
 * Output: 3 as the distance of given node 6 from the root (1).
 *
 * Node: 8
 * Output: 4
 *
 * Node: 6
 * Output: 3
 **/

'use strict';

/**
 * Node built-in dependencies
 **/

const fs = require('fs');

const path = require('path');

/**
 * Here, our Tree class with the method to find distance from root (PWC
 * Solution)
 **/

class Tree {
  constructor(root) {
    this.root = root || null;
  }
  // For building out tree
  addNode(value, parentValue) {
    if (!value) return null;
    const newNode = { children: [], value };

    if (this.root === null) {
      this.root = newNode;
      return;
    }

    this.#crawl((node) => {
      if (node.value === parentValue) {
        newNode.parent = node;
        node.children.push(newNode);
        return this;
      }
    });
  }
  // Private method to traverse node
  #crawl(callback) {
    const recurse = (node) => {
      callback(node);
      node.children.forEach((child) => {
        recurse(child);
      });
    }
    recurse(this.root);
  }
  // This method provides our solution
  distanceFromRoot(value) {
    let distance = 0;
    let node = this.find(value);

    if (!node.parent) return distance;

    while (node.parent) {
      distance += 1;
      node = node.parent;
    }
    return distance;
  }
  // To find our target node
  find(value) {
    let found = false;
    this.#crawl((node) => {
      if (node.value === value) {
        found = node;
        return;
      }
    });
    return found;
  }
}

/**
 * Followed by some utilities to test our solution
 **/

function parseTestCase(filePath = '') {
  try {
    const data = fs.readFileSync(filePath, 'utf8');

    const [lines, tests] = data.split('\n')
      .reduce(([tmpLines, tmpTests], line) => {
        if (line.charAt(0) === '#' || line.length === 0) {
          return [tmpLines, tmpTests];
        }

        if (line.split('').indexOf(',') !== -1) {
          const test = line.trim().split(',')
            .map(char => parseInt(char.trim()));

          tmpTests.push(test);
          return [tmpLines, tmpTests];
        }

        tmpLines.push(line);
        return [tmpLines, tmpTests];
      }, [[], []]
    );

    const tree = new Tree();

    lines.forEach((line, idx) => {
      if (tree.root === null) {
        const value = parseInt(line.trim());

        tree.addNode(value);
        return;
      }

      let connections = /\\|\//;
      if (line.match(connections)) return;

      const parents = lines[idx - 2].split('').filter(char => {
        if (char !== ' ') return char;
      }).map(val => {
        return tree.find(parseInt(val));
      });

      const children = line.split('').reduce((matched, char, i) => {
        if (char !== ' ') {
          matched.push([parseInt(char), i]);
        }
        return matched;
      }, []);

      children.forEach(child => {
        const [ value, i ] = child;

        const connector = lines[idx - 1]
          .split('')
          .slice(i - 1 < 0 ? 0: i - 1, i + 2)
          .reduce((parentIdx, char) => {
            if (char.match(connections)) {
              return char === '/' ? i + 2: char === '\\' ? i - 2: null;
            }
            return parentIdx;
          }, null
        );
        const parentVal = parseInt(lines[idx - 2].split('')[connector]);

        const parent = parents.reduce((match, node) => {
          if (node.value === parentVal) return node;
          return match;
        }, null);

        const newNode = { children: [], parent, value };

        parent.children.push(newNode);
      });
    });
    const display = lines.join('\n');

    return { tree, tests, display };
  } catch (error) {
    console.log('Problems parsing test case at path ', filePath, '\n', error);
  }
}

function assertMatch(tree, tests) {
  tests.forEach(test => {
    const [ value, expected ] = test;

    const result = tree.distanceFromRoot(value);

    console.log(`Distance from ${value} to root`);
    console.log(`Expected: ${expected}`);
    console.log(`Result: ${result}`);

    if (expected === result) {
      console.log('\x1b[32m%s\x1b[0m', 'Passed \u2690');
    } else {
      console.log('\x1b[31m%s\x1b[0m', 'Failed \u2715');
    }
  });
}

const isFile = (filePath) => fs.lstatSync(filePath).isFile();

const isDirectory = (filePath) => fs.lstatSync(filePath).isDirectory();

/**
 * And our test runner
 **/

(function main() {
  const testPath = process.argv[2] || '../test_cases/ch-1';

  try {
    if (isFile(testPath)) {
      const { tree, tests, display } = parseTestCase(testPath);

      console.log('Case: ', testPath);
      console.log('\n' + display + '\n');

      return assertMatch(tree, tests);
      console.log('\n\n');
    }

    if (isDirectory(testPath)) {
      fs.readdirSync(testPath).map(fileName => {
        const filePath = path.join(testPath, fileName);

        const { tree, tests, display } = parseTestCase(filePath);

        console.log('Case: ', filePath);
        console.log('\n' + display + '\n');

        assertMatch(tree, tests);
        console.log('\n\n');
      });
      return;
    }

    return 'No tests found';
  } catch (error) {
    console.log('Something went wrong: ', error);
  }
})();

File: 131/perl/ch-1.pl
#!/usr/bin/perl
# ch-1.pl

=begin summary

 * https://theweeklychallenge.org/blog/perl-weekly-challenge-131/
 *
 * Task 1 > Consecutive Arrays
 * ===========================
 *
 * You are given a sorted list of unique positive integers.
 *
 * Write a scipt to return lists of arrays where the arrays are consecutive
 * integers.
 *
 * Example 1:
 *
 * my @input = (1, 2, 3, 6, 7, 8, 9);
 *
 * my @output = ([1, 2, 3], [6, 7, 8, 9]);
 *
 * Example 2:
 *
 * my @input = (11, 12, 14, 17, 18, 19);
 *
 * my @output = ([2], [4], [6], [8]);
 *
 * Example 3:
 *
 * my @input = (2, 4, 6, 8);
 *
 * my @output = ([2], [4], [6], [8]);
 *
 * Example 4:
 *
 * my @input = (1, 2, 3, 4, 5);
 *
 * my @output = ([1, 2, 3, 4, 5]);

=end summary
=cut

use strict;
use warnings;
use utf8;
use open ":std", ":encoding(UTF-8)";
use Term::ANSIColor;
use Data::Dumper;

###################################################
# Here our Consecutive Array reducer (PWC Solution)
###################################################

sub reduce_to_consecutive {
  my @input = @{ +shift };
  my @reduced;
 
  foreach my $element (@input) {
    # if `@reduced` is empty, push the first element as part of an anonymous
    # list to start our comparisons 
    if (not (scalar @reduced)) {
      push(@reduced, [$element]);
      next;
    }
    # grab a copy of the last element checked to test if consecutive
    my $last_element = $reduced[-1]->[-1];
    # increment $last_element and compare to current element 
    if (++$last_element eq $element) {
      # if matched, push the current element to the last set
      push @{ $reduced[-1] }, $element;
      next;
    }
    # Else, push it as part of a new set.
    push(@reduced, [$element]);
  }
  return \@reduced;
}

###################################################
###################################################
#
###################################################
# Followed by some utilities to test our solution
###################################################

sub eval_input {
  my $input = shift;
  # If no inner sets, return the split string as an array reference
  if (not ($input =~ /\[/)) {
    # Strip outer parenthesis
    $input =~ s/\(|\)//g;
<<<<<<< HEAD
    # split by comma and optional spaces
    return [ map int, split /\s*,\s*/, $input];
  }
  # Else, return a reference to an array of sets
  return [
    map {
      [ map int, split /\s*,\s*/ ]
    } $input =~ /\[([^\]]*)\]/g
  ];
}

sub parse_test_case {
  # Parse file into arrays of inputs and answers
  my $file_path = shift;
  my @inputs;
  my @answers;

  open my $fh, "<", $file_path
=======
    # split by comma and optional spaces, map values to int (just to take the
    # challenge literally), return reference to anonymous list
    return [ map int, split /\s*,\s*/, $input ];
  }
  # Else, return a reference to an array of lists
  my @inputs;
  while ($input =~ /\[([^\]]*)\]/g) {
    push @inputs, [ map int, split /\s*,\s*/, $1 ];
  }
  return \@inputs;
}

sub parse_test_case {
   my $file_path = shift;
   my @inputs;
   my @answers;

   open my $fh, "<", $file_path
>>>>>>> c8dd1c6ea3275a1720568c049b13f34fa622b832
    or die "Could not open '$file_path' - $!\n";

  while (my $line = <$fh>) {
    chomp $line;
    # Skip comments
<<<<<<< HEAD
    next if $line =~ /^\s*#|^\s*$/;
    # Trim whitespace
    $line =~ s/^\s+|\s+$//g;
    # Parse line into list reference
    my $parsed = eval_input $line;
    # If there are more inputs than answers, assume line is an answer
=======
    next if $line =~ /^(\s*|\s*#.*)$/;
    # Trim whitespace
    $line =~ s/^\s+|\s+$//g;
    my $parsed = eval_input $line;
    # If inputs greater than answers, assume line is an answer
>>>>>>> c8dd1c6ea3275a1720568c049b13f34fa622b832
    if (scalar @inputs > scalar @answers) {
      push @answers, $parsed;
      next;
    }
<<<<<<< HEAD
    # Else, assume it an input
    push @inputs, $parsed;
  }
  return (\@inputs, \@answers);
}

sub assert_deep_match {
  my @first = @{ +shift };
  my @second = @{ +shift };
  my $match = 1;
  # If lengths don't match, fail
  if (scalar @first != scalar @second) {
    return 0;
  }
  # Check each element
  foreach my $idx (0 .. $#first) {
    my $ref = ref $first[$idx];
    # if element is a scalar check for match
    if ($ref ne "ARRAY" && $first[$idx] ne $second[$idx]) {
      $match = 0;
    }
    # if element is an array recurse
    if ($ref eq "ARRAY") {
      my $recursed = assert_deep_match($first[$idx], $second[$idx]);
      $match = $recursed;
    }
    # Quit early
    last if not $match;
  }
  return $match;
}

sub deep_print {
  my @input = @{ +shift };

  foreach my $idx (0 .. $#input) {
    my $el = $input[$idx];
    my $ref = ref $el;

    if ($ref eq "ARRAY") {
      print "[";
      deep_print($el);
      print "]";
    } else {
      print $el;
    }
    print ", " unless $idx == $#input;
  }
}

sub print_results {
  my $test_path = shift;
  my @inputs = @{ +shift };
  my @answers = @{ +shift };

  print $test_path, "\n";
  print "=" x length($test_path), "\n\n";

  foreach my $idx (0 .. $#inputs) {
    my $input = $inputs[$idx];
    # Check for empty array (just to be safe)
    next unless scalar @{ $input };

    my $answer = $answers[$idx];
    my $result = reduce_to_consecutive $input;

    print "Input: (", join(", ", @{ $input }), ")\n";
    print "Expected: (";
    deep_print($answer);
    print ")\n";
    print "Result: (";
    deep_print($result);
    print ")\n";

    if (assert_deep_match($answer, $result)) {
      print color("green"), "Passed \x{2690}\n", color("reset");
    } else {
      print color("red"), "Failed \x{2715}\n", color("reset");
    }
    print "\n";
  }
  print "\n";
}

###################################################
###################################################
#
###################################################
# And our test runner
###################################################

sub main {
  my $target = shift @ARGV // "../test_cases/ch-1";
  # Handle single file target
  if (-e -r -f $target) {
    my ($inputs_array, $answers_array) = parse_test_case $target;
    print_results $target, $inputs_array, $answers_array;
    return;
  }
  # Handle directory target
  if (-e -r -d _) {
    # Check for trailing slash
    $target =~ s/^(.*?)\/?$/$1\//;

    opendir my $dh, $target
      or die "Could not open '$target' - $!\n";

    my @entries = readdir $dh;
    closedir $dh;

    foreach my $entry (sort @entries) {
      # Skip the garbage
      next if $entry eq "." or $entry eq "..";
      my $path = $target . $entry;
      my ($inputs_array, $answers_array) = parse_test_case $path;
      print_results $path, $inputs_array, $answers_array;
    }
    return;
  } else {
    print "No tests found at $target\n";
  }
}

main();
=======
    # else, push to inputs
    push @inputs, $parsed;
  }
  return [\@inputs, \@answers];
}

sub assert_match {
  my (@set_1, @set_2, $flag) = ( @{ @_[0, 1] }, $_[2] );
  my $match = 1;
  print Dumper @set_1;
  print Dumper @set_2;
  print Dumper $flag;
}

my $input = '([1, 2], [3, 4], 5)';
my @arr1 = qw(1 2 3 4);
my @arr2 = qw(7 8 9);

assert_match \@arr1, \@arr2, 0;

# my $test = parse_test_case "../test_cases/ch-1/case-1.txt";

# print Dumper $test;
>>>>>>> c8dd1c6ea3275a1720568c049b13f34fa622b832
